### PR TITLE
Send refresh token to web view to fix multiple password prompts when interaction is required

### PR DESCRIPTION
**Pull request by dsaunders45459**
In the case consent or MFA is required to access a new resource ADAL SDK prompts user to re-enter the password even though the user has already entered a password and attained a valid refresh token previously. This change will enable sending the refresh token to WKWebView so that it can be used to directly show the prompt required. The client application will need to explicitly enable -[ADAuthenticationContext useRefreshTokenForWebview] to use this new feature, which has the side effect of disabling cookie auth that WKWebView may have persisted. It will use the refresh token instead. Only session cookies were used in previous implementation, so when the application closes cookies are lost. x-ms-sso-Ignore-SSO needs to have been used when acquiring the refresh token to bypass future password prompts.

**Testing**

- Tested setting useRefreshTokenForWebview to YES in the test app, then acquiring a token for OneDrive using interactive flow, I could then restart the test app and acquire a token for OneDrive using the interactive flow.

- Tested both family refresh token and refresh token without a family id work. 

- AD_PROMPT_ALWAYS and AD_FORCE_PROMPT append &prompt=login to the querystring which forces user to enter password again. We do not send the refresh token in this case for performance.

- Tested MFA scenario with claims and consent scenario by passing AD_PROMPT_AUTO after initially acquiring refresh token to https://api.office.com/discovery.

- Tested iOS/Mac using the test app.

- Tested useRefreshTokenForWebview set to NO reverts to the previous behavior without sending refresh token.

### CODE REVIEW

File: ADAL/src/public/ADAuthenticationContext.h [hunk 1]
```diff
@@ -520,6 +523,26 @@ typedef enum
...
+ @param resource The resource for whom token is needed.
+ @param clientId The client identifier
+ @param redirectUri The redirect URI according to OAuth2 protocol
+ @param promptBehavior Controls if any credentials UI will be shown.
+ @param userId An ADUserIdentifier object describing the user being authenticated
+ @param queryParams The extra query parameters will be appended to the HTTP request to the authorization endpoint. This parameter can be nil. It should be URL-encoded.
+ @param claims The claims parameter that needs to be sent to authorization endpoint. It should be URL-encoded.
+ @param completionBlock the block to execute upon completion. You can use embedded block, e.g. "^(ADAuthenticationResult res){ <your logic here> }"
+ */
+- (void)acquireTokenInteractiveWithResource:(nonnull NSString *)resource
```
Comment (oldalton) [inline ADAL/src/public/ADAuthenticationContext.h]: why was this added?

Comment (dsaunders45459) [reply ADAL/src/public/ADAuthenticationContext.h]: This method was added to improve performance for one of our scenarios. Our app already tries to silently get an access token to a resource from another code path and know that consent is required after making 2 failed requests due to FOCI. If we then try to use the "133" API passing in AD_PROMPT_AUTO, that method will attempt 2 more tries to silently get the access token (4 identical silent requests). AD_PROMPT_ALWAYS skips the silent request, but it appends "prompt=login" to the web view request forcing user to enter password. The only difference between these two API is that this one sets skip cache so that the web view will always be used.

File: ADAL/src/cache/ADResponseCacheHandler.m [hunk 2]
```diff
@@ -88,7 +88,8 @@ + (ADAuthenticationResult *)handleError:(NSError *)msidError
                                   cache:(MSIDLegacyTokenCacheAccessor *)cache
                                  params:(ADRequestParameters *)requestParams
 {
-    if (response.oauthErrorCode == MSIDErrorServerInvalidGrant && refreshToken)
+    NSString *subError = [[msidError userInfo] objectForKey:MSIDOAuthSubErrorKey];
+    if (response.oauthErrorCode == MSIDErrorServerInvalidGrant && refreshToken && (subError == nil || [subError caseInsensitiveCompare:@"consent_required"] != NSOrderedSame))
```
Comment (oldalton) [inline ADAL/src/cache/ADResponseCacheHandler.m]: Why are we not removing tokens for any invalid grant?

File: ADAL/src/request/ADAuthenticationRequest+WebRequest.m [hunk 3]
```diff
@@ -89,7 +130,13 @@ - (void)requestCode:(MSIDAuthorizationCodeCallback)completionBlock
     MSID_LOG_VERBOSE(_requestParams, @"Requesting authorization code");
     MSID_LOG_VERBOSE_PII(_requestParams, @"Requesting authorization code for resource: %@", _requestParams.resource);
 
-    [ADWebAuthController startWithRequest:_requestParams promptBehavior:_promptBehavior context:_context completion:^(MSIDWebviewResponse *response, NSError *error) {
+    NSString *refreshToken = nil;
+    if (_promptBehavior != AD_FORCE_PROMPT && _promptBehavior != AD_PROMPT_ALWAYS && [_context useRefreshTokenForWebview])
```
Comment (oldalton) [inline ADAL/src/request/ADAuthenticationRequest+WebRequest.m]: perhaps this can be simplified to _promptBehavior == AD_PROMPT_AUTO, as refresh session is same as prompt always in most cases.

File: ADAL/src/ui/ADWebAuthController.m [hunk 4]
```diff
@@ -139,6 +144,15 @@ + (void)startWithRequest:(ADRequestParameters *)requestParams
     webviewConfig.presentationType = ADAuthenticationSettings.sharedInstance.webviewPresentationStyle;
 #endif
 
+    if ([context useRefreshTokenForWebview])
+    {
+        [[webviewConfig customHeaders] setObject:@"1" forKey:ADWebAuthIgnoreSSOHeader];
```
Comment (oldalton) [inline ADAL/src/ui/ADWebAuthController.m]: should this only be set if refresh token is present?

File: ADAL/src/public/ADAuthenticationContext.h (hunk 1 above)
Comment (oldalton) [reply ADAL/src/public/ADAuthenticationContext.h]: Sounds good, resolving this one.

File: ADAL/src/cache/ADResponseCacheHandler.m (hunk 2 above)
Comment (dsaunders45459) [reply ADAL/src/cache/ADResponseCacheHandler.m]: consent_required does not imply that the refresh token is invalid for all resources, only that interactive flow is required.

File: ADAL/src/request/ADAuthenticationRequest+WebRequest.m (hunk 3 above)
Comment (dsaunders45459) [reply ADAL/src/request/ADAuthenticationRequest+WebRequest.m]: I will use your suggestion here, although the documentation for refresh session says it does not prompt in some cases where cookies are available. We don't use AD_PROMPT_REFRESH_SESSION input.

File: ADAL/src/ui/ADWebAuthController.m (hunk 4 above)
Comment (dsaunders45459) [reply ADAL/src/ui/ADWebAuthController.m]: The identity provider only accepts the refresh token if it was acquired in a flow that includes x-ms-sso-Ignore-SSO header. https://github.com/AzureAD/azure-activedirectory-library-for-cpp/blob/f8eefdc3a7e5680b5cebd6f924f9b46e31fc7834/src/cpp/src/lib/windows/WebUITokenRequest.cpp is the c++ SDK implementation, it uses this parameter on every request. We only set it if context explicitly enables it since it causes cookies to be ignored.

File: ADAL/src/cache/ADResponseCacheHandler.m (hunk 2 above)
Comment (oldalton) [reply ADAL/src/cache/ADResponseCacheHandler.m]: Can we please update this to following?

`if (response.oauthErrorCode == MSIDErrorServerInvalidGrant && refreshToken && [subError caseInsensitiveCompare:@"consent_required"] != NSOrderedSame)`

That way we only exclude `consent_required` for the time being as agreed. Your condition will also stop removing tokens for valid conditions when token should be removed. Thanks!

File: ADAL/src/request/ADAuthenticationRequest+WebRequest.m [hunk 5]
```diff
@@ -80,6 +82,45 @@ - (void)executeRequest:(NSDictionary *)request_data
      }];
 }
 
+- (NSString *)getRefreshTokenForRequest
+{
+    if (![NSString msidIsStringNilOrBlank:_refreshToken])
+    {
+        return _refreshToken;
```
Comment (oldalton) [inline ADAL/src/request/ADAuthenticationRequest+WebRequest.m]: I cannot find where this variable would get set. Can you clarify?

File: ADAL/src/ui/ADWebAuthController.m (hunk 4 above)
Comment (oldalton) [reply ADAL/src/ui/ADWebAuthController.m]: Good point, thanks for clarifying.

File: ADAL/src/cache/ADResponseCacheHandler.m (hunk 2 above)
Comment (dsaunders45459) [reply ADAL/src/cache/ADResponseCacheHandler.m]: I believe the logic is correct as is, and removes the token for any MSIDErrorServerInvalidGrant error other than consent_required. The difference between these two conditions is that I am considering the case where subError is nil. In this case nil == 0 and NSOrderedSame == 0 so it will not remove the token. I just did extra testing around this condition to confirm this is the case.

File: ADAL/src/request/ADAuthenticationRequest+WebRequest.m (hunk 5 above)
Comment (dsaunders45459) [reply ADAL/src/request/ADAuthenticationRequest+WebRequest.m]: it is set from - [ADAuthenticationRequest setRefreshToken:]. API "136" and "137". We aren't using either of these methods, but if the refresh token is already set for the request I think it is correct to use it instead of looking in the cache.
